backends/memcheck/ceed-memcheck-vector.c

```diff
@@ -174,44 +252,129 @@ static int CeedVectorRestoreArray_Memcheck(CeedVector vec) {
-  }
-  if (impl->array_owned) {
-    memcpy(impl->array_owned, impl->array, length * sizeof(CeedScalar));
-  }
+
+  // Copy back to internal buffer and sync
+  memcpy(impl->array_allocated, impl->array_writable_copy, length * sizeof(CeedScalar));
+  CeedCallBackend(CeedVectorSyncArray_Memcheck(vec, CEED_MEM_HOST));
+
+  // Invalidate writable buffer
+  for (CeedSize i = 0; i < length; i++) impl->array_writable_copy[i] = NAN;
+  CeedCallBackend(CeedFree(&impl->array_writable_copy));
+  VALGRIND_DISCARD(impl->writable_block_id);
   return CEED_ERROR_SUCCESS;
 }
 
 //------------------------------------------------------------------------------
 // Vector Restore Array Read-Only
 //------------------------------------------------------------------------------
 static int CeedVectorRestoreArrayRead_Memcheck(CeedVector vec) {
+  Ceed                 ceed;
   CeedSize             length;
   CeedVector_Memcheck *impl;
 
+  CeedCallBackend(CeedVectorGetCeed(vec, &ceed));
   CeedCallBackend(CeedVectorGetData(vec, &impl));
   CeedCallBackend(CeedVectorGetLength(vec, &length));
 
-  CeedCheck(!memcmp(impl->array, impl->array_read_only_copy, length * sizeof(CeedScalar)), CeedVectorReturnCeed(vec), CEED_ERROR_BACKEND,
-            "Array data changed while accessed in read-only mode");
+  // Verify no changes made during read-only access
+  bool is_changed = memcmp(impl->array_allocated, impl->array_read_only_copy, length * sizeof(CeedScalar));
 
+  CeedCheck(!is_changed, ceed, CEED_ERROR_BACKEND, "Array data changed while accessed in read-only mode");
+
+  // Invalidate read-only buffer
   for (CeedSize i = 0; i < length; i++) impl->array_read_only_copy[i] = NAN;
   CeedCallBackend(CeedFree(&impl->array_read_only_copy));
+  VALGRIND_DISCARD(impl->read_only_block_id);
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Take reciprocal of a vector
+//------------------------------------------------------------------------------
+static int CeedVectorReciprocal_Memcheck(CeedVector vec) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl;
+
+  CeedCallBackend(CeedVectorGetData(vec, &impl));
+  CeedCallBackend(CeedVectorGetLength(vec, &length));
+
+  for (CeedSize i = 0; i < length; i++) {
+    if (fabs(impl->array_allocated[i]) > CEED_EPSILON) impl->array_allocated[i] = 1. / impl->array_allocated[i];
+  }
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Compute x = alpha x
+//------------------------------------------------------------------------------
+static int CeedVectorScale_Memcheck(CeedVector x, CeedScalar alpha) {
+  CeedSize             length;
+  CeedVector_Memcheck *impl;
+
+  CeedCallBackend(CeedVectorGetData(x, &impl));
+  CeedCallBackend(CeedVectorGetLength(x, &length));
+
+  for (CeedSize i = 0; i < length; i++) impl->array_allocated[i] *= alpha;
+  return CEED_ERROR_SUCCESS;
```

Comment:
Should we add a NaN check here? Similar logic as other functions, but for say `VecScale(x, 0.0)` when `x` contains `inf` or `-inf`.